cmd/collector/main.go

```diff
@@ -85,10 +101,27 @@ func main() {
 	r := mux.NewRouter()
 	apiHandler := app.NewAPIHandler(jaegerBatchesHandler, zipkinSpansHandler)
 	apiHandler.RegisterRoutes(r)
+
 	httpPortStr := ":" + strconv.Itoa(*builder.CollectorHTTPPort)
 	recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-	logger.Info("Listening for HTTP traffic", zap.Int("http-port", *builder.CollectorHTTPPort))
-	if err := http.ListenAndServe(httpPortStr, recoveryHandler(r)); err != nil {
-		logger.Fatal("Could not launch service", zap.Error(err))
+	logger.Info("Starting Jaeger Collector HTTP server", zap.Int("http-port", *builder.CollectorHTTPPort))
+
+	go func() {
+		// there's still a chance that a request comes in between the `ready` flag and the server's
+		// actual startup, but this would be a rare situation, as we assume the `ListenAndServe` is fast enough
+		// between allocating a port and serving a request.
+		hcChannel <- http.StatusNoContent
+
+		logger.Fatal("Could not launch service", zap.Error(http.ListenAndServe(httpPortStr, recoveryHandler(r))))
+
+		// if we returned from the call above, it's because we have troubles... the logger.Fatal should panic,
+		// so, this code probably won't be reached, but if we are hung for some reason, we can still notify our
+		// state to an external actor
+		hcChannel <- http.StatusInternalServerError
+	}()
+
+	select {
+	case <-serverChannel:
+		logger.Info("Jaeger Collector is finishing")
```

Comment:
I'm not sure if you're still working on this part (let me know if you are and you can ignore me) but with the current way this select is done it will receive the signal and main will instantly exit, preventing the safe shutdown. Adding a waitgroup is the most common solution in this approach.